core/src/main/java/be/orbinson/aem/dictionarytranslator/services/DictionaryService.java

```diff
@@ -1,43 +1,143 @@
 
     void createDictionary(Resource parent, String name, String[] languages, String basename) throws PersistenceException;
 
-    void deleteDictionary(ResourceResolver resourceResolver, String dictionaryPath) throws DictionaryException;
+    void deleteDictionary(ResourceResolver resourceResolver, String dictionaryPath) throws DictionaryException, ReplicationException, PersistenceException;
 
     List<String> getLanguages(Resource dictionaryResource);
 
-    void deleteLanguage(ResourceResolver resourceResolver, Resource dictionaryResource, String language) throws DictionaryException;
+    void deleteLanguage(Resource dictionaryResource, String language) throws DictionaryException, ReplicationException, PersistenceException;
 
     void addLanguage(Resource dictionaryResource, String language, String basename) throws PersistenceException;
 
-    Resource getLanguageResource(Resource dictionaryResource, String language);
-
     String getBasename(Resource dictionaryResource);
 
-    List<String> getKeys(Resource dictionaryResource);
-
-    boolean keyExists(Resource dictionaryResource, String language, String key);
-
-    Resource getMessageEntryResource(Resource languageResource, String key);
-
-    void createMessageEntry(ResourceResolver resourceResolver, Resource dictionaryResource, String language, String key, String message) throws PersistenceException;
-
-    void updateMessageEntry(ResourceResolver resourceResolver, Resource dictionaryResource, String language, String key, String message) throws PersistenceException, RepositoryException;
+    List<String> getKeys(Resource dictionaryResource) throws DictionaryException;
+
+    boolean keyExists(Resource dictionaryResource, String language, String key) throws DictionaryException;
+
+    /**
+     * Either creates a new message entry or updates an existing one. The changes are not persisted until the resource resolver is committed.
+     * @param dictionaryResource
+     * @param language
+     * @param key
+     * @param message
+     * @throws PersistenceException
+     * @throws DictionaryException 
+     */
+    void createOrUpdateMessageEntry(Resource dictionaryResource, String language, String key, String message) throws PersistenceException, DictionaryException;
+
+    /**
+     * Deletes the message entry for the given language and key. The changes are not persisted until the resource resolver is committed.
+     * However the message entry is immediately scheduled for deactivation.
+     * @param dictionaryResource
+     * @param language
+     * @param key
+     * @throws PersistenceException
+     * @throws ReplicationException
+     * @throws DictionaryException 
+     */
+    void deleteMessageEntry(Resource dictionaryResource, String language, String key)  throws PersistenceException, ReplicationException, DictionaryException;
+
+    Map<String, Message> getMessages(Resource dictionaryResource, String language) throws DictionaryException;
```

Comment:
Can we add a reason on when this would throw a dictionary exception?

Reply:
Done.